File: 2015/05/2.cpp
#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <map>
#include <string>
#include <tuple>
int main() {
  auto input = std::ifstream("./2015/05/input.txt");
  if (!input.good()) {
    std::cerr << "Error opening file './2015/04/input.txt'" << std::endl;
    return EXIT_FAILURE;
  }
  uint16_t niceCount = 0;
  std::string line;
  while (!input.eof()) {
    std::map<std::tuple<char, char>, size_t> pairs = {};
    input >> line;
    const size_t size = line.size();
    bool condition1 = false;
    bool condition2 = false;
    char prev2;
    char *pair = line.data();
    for (size_t index = 0; index + 1 < size; index++, prev2 = pair[0]) {
      pair = &((char *)line.data())[index];
      const auto tup = std::make_tuple(pair[0], pair[1]);
      const auto &f = pairs.find(tup);
      if (!condition1 && f != pairs.end() && f->second + 2 <= index) {
        condition1 = true;
      }
      if (!condition2 && (pair[1] == prev2) && (index != 0)) {
        condition2 = true;
      }
      if (condition1 && condition2) {
        goto nice;
      }
      if (!pairs.contains(tup))
        pairs[tup] = index;
    }
  naughtly:
    continue;
  nice:
    niceCount++;
  }
  std::cout << niceCount;
  return EXIT_SUCCESS;
}

File: 2015/05/2.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, err := os.Open("./2015/05/input.txt")
	if err != nil {
		fmt.Fprintln(os.Stderr, "Unable to open input file")
		os.Exit(1)
	}
	stat, _ := input.Stat()
	b := make([]byte, stat.Size())
	input.Read(b)
	niceCount := 0
	for _, line := range strings.Split(string(b), "\n") {
		pairs := make(map[string]int)
		oldpair := ""
		condition1, condition2 := false, false
		for i := 0; i+1 < len(line); i++ {
			pair := line[i : i+2]
			value, contains := pairs[pair]
			if !condition1 && contains && value+2 <= i {
				condition1 = true
			}
			if !condition2 && len(oldpair) != 0 && oldpair[0] == pair[1] {
				condition2 = true
			}
			if condition1 && condition2 {
				fmt.Println(line)
				niceCount++
				break
			}
			oldpair = pair
			if !contains {
				pairs[pair] = i
			}
		}
	}
	fmt.Println(niceCount)
}
